server/src/main/java/org/apache/druid/server/coordinator/rules/Rules.java

```diff
@@ -43,4 +48,40 @@ public static boolean eligibleForLoad(Period period, Interval interval, DateTime
   private Rules()
   {
   }
+
+  /**
+   * Validate rules. This method throws an exception if a rule contain an interval that
+   * will fully cover the next rules' interval in the list. Rules that will be evaluated at some point
+   * are considered to be legitimate.
+   * @param rules Datasource rules.
+   */
+  public static void validateRules(final List<Rule> rules)
+  {
+    if (rules == null) {
+      return;
+    }
+    final DateTime now = DateTimes.nowUtc();
+    for (int i = 0; i < rules.size() - 1; i++) {
+      final Rule currRule = rules.get(i);
+      final Rule nextRule = rules.get(i + 1);
+      final Interval currInterval = currRule.getEligibleInterval(now);
+      final Interval nextInterval = nextRule.getEligibleInterval(now);
+      if (currInterval.contains(nextInterval)) {
+        // If the current rule has an eternity interval, it covers everything following it.
+        // Or if the current rule still covers the next rule at the current interval boundaries, then the
+        // next rule will never fire at any time, so throw an exception.
+        if (Intervals.ETERNITY.equals(currInterval) ||
+            (currRule.getEligibleInterval(currInterval.getStart()).contains(nextRule.getEligibleInterval(currInterval.getStart()))
+             && currRule.getEligibleInterval(currInterval.getEnd()).contains(nextRule.getEligibleInterval(currInterval.getEnd())))) {
+          throw InvalidInput.exception(
+              "Rule[%s] has an interval that contains interval for rule[%s]. The interval[%s] also covers interval[%s].",
+              currRule,
+              nextRule,
+              currInterval,
+              nextInterval
```

Comment:
What's the behaviour in the following situation:
Rule#1 - Covers the interval 2020-2021
Rule#2 - Covers the interval 2021-2022
Rule#3 - Covers the interval 2020-2022

In this case `validateRules` would not throw an exception, however, from my understanding, Rule#3 would never trigger. 

If that's the case, we should handle these cases as well (probably by maintaining a set of intervals that the rules seen so far cover and checking if the next rule isn't completely overshadowed by the interval set seen so far)

Reply:
That sounds like a nice check to have as well. Let me think about it more and adjust as needed.

Reply:
@LakshSingla, actually, Druid currently does a fully [contains](https://github.com/apache/druid/blob/master/server/src/main/java/org/apache/druid/server/coordinator/rules/IntervalDropRule.java#L66) interval check for drop rules, so rule #3 is valid with `dropByInterval`. So we can't really guarantee the overlapping condition in a type agnostic way (which apply to load rules, so we would probably need a separate contract)

Reply:
I am sorry, I didn't get why that changes why we in the original example, rule#1 spans from [2020-2021). rule #2 spans from [2021-2022). Therefore, collectively, rule#1 and rule#2 span from [2020-2022). Is there any instant in the timeline that would cause rule#3 to fire?

Reply:
From my understanding, rule #3 is valid for the time interval if it existed in isolation. However, when present along with other rules, the precedence of those takes place, causing rule#3 to be effectively useless.

Reply:
Yes, I agree. Rule #3 would never be applied on any segment irrespective of the rule types.
While evaluating retention rules, as soon as a rule applies to a segment, we break and don't look at the other rules.

Also, in general, I am not entirely sure of the validation logic here. 
- For one, the set thing suggested by @LakshSingla makes sense. But we would have to ensure that the `fully covered` check is done correctly.
  - Additive logic (cleaner to implement): Say your set contains intervals `A [4, 10)` and `B [15, 20)` and a new interval `C [8, 17)` comes in. Adding this new interval C should merge all of A, B and C into one interval `[4, 17)`.
  - Subtractive logic: To check if an interval has any part which is not fully covered by the preceding intervals, remove all parts of it that overlap with intervals already encountered. If you are left with an empty interval, the rule is not valid.
- Secondly, all of this is easy to follow when we are dealing with `ForeverXyzRule` or `IntervalXyzRule` (rule with fixed interval). For `PeriodXyzRule`, I am not entirely sure of the boundaries that would need to be checked. The proposed change does compute the `eligibleInterval` at both start and end of the `currInterval` but I need to give it some more thought.
- There could even be other types of rules in the future, but I guess we need not worry about them now.

Reply:
Yeah, let's consider a segment with an interval `2020/2022` and the following drop rules in order:

1. `dropByInterval('2020/2021')`
2. `dropByInterval('2021/2022')`
3. `dropByInterval('2020/2022')`

For the segment in question, only rule 3 will match as the _complete_ interval is [contained](https://github.com/apache/druid/blob/master/server/src/main/java/org/apache/druid/server/coordinator/rules/IntervalDropRule.java#L66) in the drop interval. 

On the other hand, if we consider `loadByInterval` rules for the same intervals, you're right that all the rules will apply for the segment as load rules will fire as long as there is at least an [overlap](https://github.com/apache/druid/blob/master/server/src/main/java/org/apache/druid/server/coordinator/rules/Rules.java#L30). 

Hope that clarifies why we cannot detect the overlapping case without baking in some type awareness, but let me know if I am missing something!

Reply:
I get the incorrect assumption I made in my original example. This digression occurs from the semantics of how intervals are applied in load rules v/s in drop rules. In load rules, we only need for partial overlap for the rule to fire, while in the drop rule, we need a complete overlap for the rule to fire

 However when we are doing a `contains` check in this logic, we are inherently assuming something about the structure of the rules. 

Consider the following case:
The system has 2 types of segments, one ranging from 2020-2021 and the other from 2020-2022. If the user wants to drop the segments ranging from 2020-2021 while preserving the other segments, how does the user craft intuitive rules for the same without causing the newly added check to fail? I only came up with the following, however, that fails the check. 
1.  Drop 2020-2021
2. Load 2020-2022

I think this ties back to the common use case that we are trying to prevent the user from doing?

Reply:
@kfaraz - please see my responses to your comments:

> we would have to ensure that the fully covered check is done correctly.

To check if an interval fully contains another, the proposed algorithm has a nested loop that goes through the sequence of rules to find if a larger interval contains another subsequent interval down in the list; O(n^2) algorithmic runtime, but practical runtime should be trivial though.

I suppose there are alternate approaches like condensing intervals - the additive and subtractive logic you're describing. My first thought is that this could work well for the overlapping case (which isn't the intent of this patch).

Re this:

> Secondly, all of this is easy to follow when we are dealing with ForeverXyzRule or IntervalXyzRule (rule with fixed interval). For PeriodXyzRule, I am not entirely sure of the boundaries that would need to be checked. The proposed change does compute the eligibleInterval at both start and end of the currInterval but I need to give it some more thought.

The boundary check primarily exists to allow the case where a rule will fire at some point. Unlike the forever and interval rules, period rules factor in the reference timestamp thereby yielding distinct intervals at the boundaries. Something like:

`PeriodRuleR3.eligibleInterval(now) = IntervalR3`
`PeriodRuleR3.eligibleInterval(1yAgo) = IntervalR4`

The boundaries check is essentially to verify that a legit rule will fire at some point and allow such a rule. For example, consider the rules `[loadByPeriod(P10Y)`,  `loadByInterval('2020/2025')`]:
- With the boundary check (before and after 10 years from now), the first rule will _not_ contain the second interval.  So the validation check will not flag the rule set. The second rule will indeed fire at some point in the future.
- Without the boundary check, the first rule will fully cover the second rule's interval, incorrectly flagging the second rule.

I realize I need to update the javadocs since it's not pretty straightforward to conceptualize this. :)